### DISCUSSION TITLE
How to block recording bots using axios/0.21.4?

### DISCUSSION THREAD

**Discussion by radio2019** [AzuraCast/AzuraCast#5047] [Q&A] [answered]
Is there any way to block axios/0.21.4?
I have an extremely large number of recording bots on my web radio stream that use this software: https://prnt.sc/26lxkau
I tried blocking the IPs, but it doesn't work as there are tons of IPs in use, all from Google ranges.
In all cases axios/0.21.4 is used, so the most effective way to solve these problems would be to block this client.
I'm currently wasting a lot of traffic on my VPS just for these bots.

**Comment by ItsMitchh** [comment]
Hi.

While we don't currently offer a native way of blocking specific user-agents, you can customize the Nginx configuration to block the user-agents, **But we will not give any support on this, only do this if you know what you're doing**. 

We know this can be highly frustrating to users and server developers, please be patient while I review some options on what we can do.

**Reply by Catskill909** [reply]
See this all the time. I ban the IPs but new ones pop up. 
The worst culprits are in a similar IP range and always appear in same location, the middle of the Thames in London in the reports map :)

**Comment by ivellios1988** [comment]
Isn't it likely that they will change their user agent once they find out they got banned?

**Comment by xewl** [comment]
Hm.. axios is a promise-based javascript library, how are they even using it to "listen" to the endpoint itself? (I'm wondering if that can even be done..)

**Reply by Vaalyn** [reply]
Well, webradio streams from Icecast are just infinite HTTP GET downloads so if you "download" the stream via axios and output or save each incoming chunk of data it surely can be used to listen to a stream. The user agent itself can definitely be changed since it's just a normal request header and axios supports setting custom headers. Btw, axios is also one of the JS libraries that are used by AzuraCast itself (not to listen to the stream though).

**Comment by YosefAdPro** [comment]
Try adding in
Station Profile > Custom Configuration

`<deny-agents>axios</deny-agents> 
`

**Reply by a11ymit** [reply]
Have you tried this yourself by any chance? I've just banned 40+ IPs I found in listener reports.

Otherwise I will have to try this tomorrow when a new IP connects so I can confirm it works.

**Reply by YosefAdPro** [reply]
Unfortunately, I have it too.

![image](https://user-images.githubusercontent.com/11810187/153381662-9bfb771e-95b2-4f05-b87a-3c26bdbddee5.png)

I will try it at night as it requires Restart to icecast.

**Reply by a11ymit** [reply]
OK cool.

If only this update was in a stable release then we wouldn't have to wait 😂

![Screenshot_20220210-011115_Chrome.jpg](https://user-images.githubusercontent.com/56458484/153382599-5d40e8f7-a634-4bb7-9311-be8159e45348.jpg)

**Reply by Catskill909** [reply]
Did this work for anyone who tried?

> Try adding in Station Profile > Custom Configuration
> 
> `<deny-agents>axios</deny-agents> `

**Reply by YosefAdPro** [reply]
As of the latest rolling-release we now support managing the User Agent ban list directly via the station profile, which will write to the "deny-agents" list.

**Comment by radio2019** [comment]
Unfortunately, bannimg those IPs does not really help (I tried it myself):
I still get connections from those IPs, however for a shorter period of
time.
I tried banning all listeners from Finland; that helped, but it also means
that regular listener from that country are unable to connect. Furthermore,
only few of those IPs are assigned to Finland; most of them are geolocated
to Great Britain where I do have some real listeners I don't want to ban.
Sadly AzuraCast doesn't allow blocking certain user agents, that would very
likely solve all those issues.

allymitton ***@***.***> schrieb am Do., 10. Feb. 2022, 03:06:

**Reply by a11ymit** [reply]
Yes I understand it won't work as they have a range of IP addresses to connect from. I have also had genuine listenes from Helsinki so that rules out a geo ban. I've noticed connections from UK IP addresses too.

That's why I'm asking if the user agent block works as I haven't seen anything else use that HTTP client other than this unknown organisation. Banning this agent is a great solution to an annoying problem.

**Comment by radio2019** [comment]
Nice to read that I'm not the only person having that issue.
If it's possible to ban certain IPs or locations, I wonder why it should be
impossible to ban certain clients.
AzuraCast should implement that feature in future releases.

allymitton ***@***.***> schrieb am Do., 10. Feb. 2022, 10:46:

**Comment by BusterNeece** [comment, marked as answer]
As of the latest rolling-release we now support managing the User Agent ban list directly via the station profile, which will write to the "deny-agents" list. This applies only to Icecast users, as no such feature exists on Shoutcast.

**Reply by YosefAdPro** [reply]
@SlvrEagle23 This feature is also present in Shoutcast. It can be managed through the Shoutcast interface.
![image](https://user-images.githubusercontent.com/11810187/153497424-26babce3-f03d-4079-a121-70b3f87fea75.png)

It produces a file named  `sc_serv.agent` under `/var/azuracast/stations/station_name/config`

**Reply by BusterNeece** [reply]
@YosefAdPro Does it update the `sc_serv.conf` file?

**Reply by YosefAdPro** [reply]
Yes.
![image](https://user-images.githubusercontent.com/11810187/153498522-534cff25-8278-48db-95c0-ec2fc6e7d595.png)

**Reply by BusterNeece** [reply]
@YosefAdPro Re-read my question...does it update the `sc_serv.conf` file itself, as in the actual main Shoutcast config, not the agents file.

**Reply by Vaalyn** [reply]
It does not (at least on my test server):

<img width="170" alt="image" src="https://user-images.githubusercontent.com/13745863/153498854-66c46b9a-ea1f-4fb9-92ea-a4c45a4c0af1.png">

<img width="80" alt="image" src="https://user-images.githubusercontent.com/13745863/153498921-3a870319-085e-4a26-94c8-670ad1b5ed7c.png">

**Reply by BusterNeece** [reply]
Looking at the Shoutcast docs, it appears there's an `agentfile` option we can provide that would mimic the IP ban format.

Pushed that in the latest rolling-release version that's currently deploying (and should be available shortly).

**Reply by Vaalyn** [reply]
@SlvrEagle23 Which docs are you looking at?

I'm somehow too blind to find anything regarding this in the official shoutcast.com docs nor in their wiki.

The only reference I've seen was on [shoutca.st](https://sc-mirror.shoutca.st/docs/DNAS_Server.html#UserAgents) for 2.5.5 from 2017:

> blockemptyuseragent : Specify if the server will allow listener connections without a user agent or not. [Default = 0]
> agentfile : File to store the list of blocked user agents [Default = sc_serv.agent]
> saveagentlistonexit : Write any changes to the 'agentfile' when required [Default = 1]

**Reply by BusterNeece** [reply]
@Vaalyn That's the same source I'm referencing. Hopefully it works.

**Reply by Vaalyn** [reply]
😆 typical of big corp software. Have some FAQs but updating their own config file reference in their wiki? No, why should they...

**Comment by Catskill909** [comment]
I've been using streaming services for a long time in many different formats and this issue of clients who latch on to record has been around for awhile.
Right now seems to be just the London and Finland similar ips and same locations on the listener report maps. So just one or two culprits do this?
What I don't understand is why? To build a monster music catalog? Music is everywhere now. Why do they spend all the time, energy and computing power doing this?
Could they just be hijacking the stream in order to place ads in some apps or something like that?

**Comment by radio2019** [comment]
That's something I ask myself, too.

Catskill909 ***@***.***> schrieb am Sa., 12. Feb. 2022, 02:38:

**Comment by radio2019** [comment]
I tried with Dalvik/2.1.0 for testing purposes, but it didn't work: The
stream is still accessible with XiiaLive.

Catskill909 ***@***.***> schrieb am So., 13. Feb. 2022, 20:05:

**Comment by BusterNeece** [comment]
We've implemented this into our Rolling Release version, which will shortly be included in our stable release. If you're still having issues with this feature *and you're on the latest Rolling Release version*, please feel free to create a new GitHub issue.
